[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33719 [profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()**
getEventList.record()**
getEventList.record()**

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

To benchmark the performance we have added a simple python script that runs the profiler over a bunch of iterations of a tight loop that uses the profiler. Here are the numbers before/after the change with 1000 iterations:

before: `Iters: 1000 Profiler Latency: 0.38266039061546325 s`
after: `Iters: 1000 Profiler Latency: 0.3876141834259033 s`

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

rohan-varma committed:
[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()

As title

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

albanD reviewed (commented):
Is there any performance impact of this change wrt to the profiler overhead?

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

rohan-varma commented:
> Is there any performance impact of this change wrt to the profiler overhead?

@albanD in the general case without RPC, this just adds an extra lock of an unlocked mutex, which isn't very expensive.

For RPC profiling, I can potentially see a case where several concurrent RPC calls attempt to write to the same `RangeEventList`, and this causes some lock contention. Though, in this case, it would be important to serialize the writes, so I'm not sure if we have a better choice here.

albanD commented on torch/csrc/autograd/profiler.h:
nit: we usually call them "mutex_" in the autograd subfolder if they don't have a special name. "mut" looks a bit weird.

albanD reviewed (commented):
Sounds good for the perf.
I guess the current use case was single threaded so taking one lock should not be a problem.
In the multi-threaded case, there is no way around this.

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

rohan-varma commented on torch/csrc/autograd/profiler.h:
I changed it to mutex_ in the update.

rohan-varma commented on torch/csrc/autograd/profiler.cpp:
This isn't related though it makes debugging a lot easier when one is logging the `Events` in the cpp code (because it records the name of the op in the event). For example I had to log them to determine why they don't match up when investigating this bug. Although, it's not strictly necessary and I can revert these changes if you prefer that.

albanD commented on torch/csrc/autograd/profiler.cpp:
For future PR, that would be good to have them as a different PR in the stack. You can leave it here for this one, no problem.

albanD reviewed (commented):
Why does the `allocBlock()` function does not need to protect its use of `blocks`? It is protected somewhere else?

rohan-varma commented:
@albanD the only call to `allocBlock` is in the function `record`, and we hold the mutex before calling it. Essentially I want to hold the mutex throughout the critical section, which is writing to `blocks` and possibly allocating if needed. Were you referring to something else?

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

albanD commented:
Right, maybe add a comment next to that function to warn that it is assumed that the mutex is held when that function is called.

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/autograd/profiler.h:
if allocBlock is only called in record, and we assume that lock is held, shall we make it in private?

mrshenli commented on torch/csrc/autograd/profiler.h:
question, given we added a lock here, will it be faster if we 1) swap/move the blocks, 2) unlock, 3) do the insertion? Looks like `std::forward_list` does have a O(1) move constructor.

> 7) Move constructor. Constructs the container with the contents of other using move semantics. Allocator is obtained by move-construction from the allocator belonging to other.

mrshenli reviewed (approved):
LGTM! Left some minor comments.

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct
getEventList.record()**
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* #33030 [profiler] add RecordFunctionAsync struct

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

rohan-varma commented on torch/csrc/autograd/profiler.h:
I think it's just an API inconsistency and we should try to use the same in both. I will send a follow up diff.

rohan-varma commented on torch/csrc/autograd/profiler.h:
It looks like we call blocks.clear() at the end as well. I think if we move from blocks, then blocks.clear() will be undefined? Although, I think `swap` should work since we can swap it with an empty forward_list.

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

ilia-cher commented on torch/csrc/autograd/profiler.cpp:
could you please split this into a separate PR? I think it doesn't belong in this PR and it should be very easy to do

ilia-cher commented on torch/csrc/autograd/profiler.h:
also, as mentioned above, you will have an opportunity to fix this in a separate PR

ilia-cher commented on torch/csrc/autograd/profiler.h:
this change is quite serious, if you make RangeEventList thread safe, then I guess there's no point in using thread_local event lists in profiler at all? I guess we should either keep it lock-free (and then update profiler's code to account for that, i.e. we can no long have both thread local and not thread local accesses, e.g. we can have internally two event lists, one is fast (lock-free) and only for thread local access, another is locked and slow, e.g. for cases like async execution that calls end() callbacks on a different thread); Or, alternatively, we could make RangeEventList thread local and simplify code, but then we need to benchmark this carefully to ensure this doesn't cause regression

ilia-cher commented on torch/csrc/autograd/profiler.h:
if you could add a small benchmark (e.g. profiling a tight loop with a recorded op inside) and show there's no regression (it might be actually fine since this lock is not contended, assuming most of the time event list is accessed by the same thread), then the current approach should be fine

rohan-varma commented on torch/csrc/autograd/profiler.h:
Discussed it a bit more offline and we will add a benchmark to profile the cost of adding a lock.

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"


We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**
getEventList.record()**

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

This PR also makes a small change to pass in the `RecordFunction` name into `popRange`. It makes the debugging easier when investigating the events recorded.

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

ilia-cher reviewed (approved):
thanks for adding benchmark!

ilia-cher commented on benchmarks/profiler_benchmark/profiler_bench.py:
btw no need to manually use timers, we can use python's timeit package

rohan-varma commented on benchmarks/profiler_benchmark/profiler_bench.py:
Sounds good, I'll update the PR.

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**


We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

To benchmark the performance we have added a simple python script that runs the profiler over a bunch of iterations of a tight loop that uses the profiler. Here are the numbers before/after the change with 1000 iterations:

before: `Iters: 1000 Profiler Latency: 0.38266039061546325 s`
after: `Iters: 1000 Profiler Latency: 0.3876141834259033 s`

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler][rpc] fix a race condition in the profiler when multiple threads call
getEventList.record()"

getEventList.record()**
getEventList.record()**

We were seeing a strange error where gathering profiler events (specifically `parse_cpu_trace` in `profiler.py`) would fail with the error:
`IndexError: pop from empty list`.

It turned out that this was because for one particular `Event`, there was a pop recorded but not a push. Instead of the `push` event being completely missing, it was overwritten by a completely different event.

After a bunch of debugging, and trying several hypotheses, it turns out that this was a race condition in `RangeEventList::record`. What happened was that different threads would call into `RangeEventList::record` on the same event list instance, and one record would stomp over the data written by the other one. Somehow the data written was a valid `Event` so the error did not manifest itself until the profiler realized a `pop` was missing a matching `push` in the python code.

I fixed this by adding a lock to serialize writes to `RangeEventList::record`.

To benchmark the performance we have added a simple python script that runs the profiler over a bunch of iterations of a tight loop that uses the profiler. Here are the numbers before/after the change with 1000 iterations:

before: `Iters: 1000 Profiler Latency: 0.38266039061546325 s`
after: `Iters: 1000 Profiler Latency: 0.3876141834259033 s`

Differential Revision: [D20071125](https://our.internmc.facebook.com/intern/diff/D20071125/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20071125/)!

[ghstack-poisoned]
